pkg/apiserver/resthandler.go

```diff
@@ -171,6 +172,85 @@ func CreateResource(r RESTCreater, ctxFn ContextFunc, namer ScopeNamer, codec ru
 	}
 }
 
+// PatchResource returns a function that will handle a resource patch
+func PatchResource(r RESTPatcher, ctxFn ContextFunc, namer ScopeNamer, codec runtime.Codec, resource string, admit admission.Interface) restful.RouteFunction {
+	return func(req *restful.Request, res *restful.Response) {
+		w := res.ResponseWriter
+
+		// TODO: we either want to remove timeout or document it (if we document, move timeout out of this function and declare it in api_installer)
+		timeout := parseTimeout(req.Request.URL.Query().Get("timeout"))
+
+		namespace, name, err := namer.Name(req)
+		if err != nil {
+			notFound(w, req.Request)
+			return
+		}
+		ctx := ctxFn(req)
+		ctx = api.WithNamespace(ctx, namespace)
+
+		original, err := r.Get(ctx, name)
```

Comment:
----- Original Message -----
> > +func PatchResource(r RESTPatcher, ctxFn ContextFunc, namer ScopeNamer,
> > codec runtime.Codec, resource string, admit admission.Interface)
> > restful.RouteFunction {
> > +	return func(req *restful.Request, res *restful.Response) {
> > +		w := res.ResponseWriter
> > +
> > +		// TODO: we either want to remove timeout or document it (if we
> > document, move timeout out of this function and declare it in
> > api_installer)
> > +		timeout := parseTimeout(req.Request.URL.Query().Get("timeout"))
> > +
> > +		namespace, name, err := namer.Name(req)
> > +		if err != nil {
> > +			notFound(w, req.Request)
> > +			return
> > +		}
> > +		ctx := ctxFn(req)
> > +		ctx = api.WithNamespace(ctx, namespace)
> > +
> > +		original, err := r.Get(ctx, name)
> 
> Makes sense. We could validate that the client passes a resource version in
> their patch. The intent was to add PATCH without needing to add an extra
> Patch method to the registry's RESTs. What's the reason that the registry
> does a get from etcd after update?

Get before update - it's so that we know that what is stored so we can do a CAS.  In the pod changes I made the REST logic happen between the initial get and the CAS, so that you had the existing state to compare to.  There's a problem there which is merging - when we merge what the server has with the client's input, we have to make the decision at that time about what to preserve (that's why pod/status was going to exist, so it could have a different decision).  Right now we throw away the status section and combine it with the spec (so that users who naively GET -> PUT overwrite don't overwrite status).  If we didn't do that, all clients would have to do merges correctly (which they aren't) and some users *should* be prevented from changing status like pod.status.host, which they aren't today.  Admission control probably would be better to run given both the old and the new value as well, and being able to calculate exact deltas for authorization is a nice to have.

Let's slate a time to sit down and go through all the edge cases here - I like the approach, it just highlights how we're not doing everything right for Update today.

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4706/files#r25143674